File: Data Structures & Algorithms/islands-and-treasure/submission-2.cpp
class Solution {
public:
    void islandsAndTreasure(vector<vector<int>>& grid) {
        // find all treasure chest
        queue<pair<int, int>> q;
        int row = grid.size(), col = grid[0].size();
        for(int i=0; i<row; i++){
            for(int j=0; j<col; j++){
                if(grid[i][j] == 0) q.push({i, j});
            }
        }

        vector<pair<int, int>> dir = {
            {0, 1},
            {0, -1},
            {1,0},
            {-1, 0}
        };

        while(!q.empty()){
            // update its neighbours
            auto [i,j] = q.front(); q.pop();
            for(auto [x, y] : dir){
                int newi = i + x;
                int newj = j + y;

                // Check if its valid grid
                if(newi < 0 || newj < 0 || newi == row || newj == col) continue;

                int val = grid[newi][newj];
                if(val == -1 || val != INT_MAX) continue; // water or it already been updated before

                // update grid and queue it
                grid[newi][newj] = grid[i][j] + 1;
                q.push({newi, newj});
            }
        }

    }
};

File: Data Structures & Algorithms/search-2d-matrix/submission-2.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        ROW, COL = len(matrix), len(matrix[0])
        i, j = 0, COL-1

        while i >= 0 and j >= 0 and i < ROW and j < COL:
            if matrix[i][j] == target:
                return True

            if matrix[i][j] > target:
                j -= 1
            else:
                i += 1
        
        return False